homeassistant/components/yamaha/media_player.py

```diff
@@ -79,78 +77,94 @@
 )
 
 
-def setup_platform(hass, config, add_entities, discovery_info=None):
+class YamahaConfigInfo:
+    """Configuration Info for Yamaha Receivers."""
+
+    def __init__(self, config: None, discovery_info: None):
+        """Initialize the Configuration Info for Yamaha Receiver."""
+        self.name = config.get(CONF_NAME)
+        self.host = config.get(CONF_HOST)
+        self.ctrl_url = f"http://{self.host}:80/YamahaRemoteControl/ctrl"
+        self.source_ignore = config.get(CONF_SOURCE_IGNORE)
+        self.source_names = config.get(CONF_SOURCE_NAMES)
+        self.zone_ignore = config.get(CONF_ZONE_IGNORE)
+        self.zone_names = config.get(CONF_ZONE_NAMES)
+        self.from_discovery = False
+        if discovery_info is not None:
+            self.name = discovery_info.get("name")
+            self.model = discovery_info.get("model_name")
+            self.ctrl_url = discovery_info.get("control_url")
+            self.desc_url = discovery_info.get("description_url")
+            self.zone_ignore = []
+            self.from_discovery = True
+
+
+async def async_setup_platform(hass, config, async_add_entities, discovery_info=None):
     """Set up the Yamaha platform."""
 
     # Keep track of configured receivers so that we don't end up
     # discovering a receiver dynamically that we have static config
     # for. Map each device from its zone_id to an instance since
     # YamahaDevice is not hashable (thus not possible to add to a set).
-    if hass.data.get(DATA_YAMAHA) is None:
-        hass.data[DATA_YAMAHA] = {}
-
-    name = config.get(CONF_NAME)
-    host = config.get(CONF_HOST)
-    source_ignore = config.get(CONF_SOURCE_IGNORE)
-    source_names = config.get(CONF_SOURCE_NAMES)
-    zone_ignore = config.get(CONF_ZONE_IGNORE)
-    zone_names = config.get(CONF_ZONE_NAMES)
-
-    if discovery_info is not None:
-        name = discovery_info.get("name")
-        model = discovery_info.get("model_name")
-        ctrl_url = discovery_info.get("control_url")
-        desc_url = discovery_info.get("description_url")
-        receivers = rxv.RXV(
-            ctrl_url, model_name=model, friendly_name=name, unit_desc_url=desc_url
-        ).zone_controllers()
-        _LOGGER.debug("Receivers: %s", receivers)
-        # when we are dynamically discovered config is empty
-        zone_ignore = []
-    elif host is None:
-        receivers = []
-        for recv in rxv.find():
-            receivers.extend(recv.zone_controllers())
-    else:
-        ctrl_url = f"http://{host}:80/YamahaRemoteControl/ctrl"
-        receivers = rxv.RXV(ctrl_url, name).zone_controllers()
-
-    devices = []
+    known_zones = hass.data.setdefault(DATA_YAMAHA, set())
+
+    config_info = YamahaConfigInfo(config=config, discovery_info=discovery_info)
+
+    def _discovery(config_info):
```

Comment:
We can move the discovery function outside `async_setup_platform` now after introducing the config_info parameter. Nice!